compiler/resolution/visibleFunctions.cpp

```diff
@@ -645,21 +615,15 @@ void getVisibleFunctions(const char*      name,
                           visited, visibleFns, false);
 }
 
-static BlockStmt* getVisibleFnsInstantiationPt(BlockStmt*    block,
-                                               ModuleSymbol* inMod,
-                                               FnSymbol*     inFn) {
+static BlockStmt* getVisibleFnsInstantiationPt(BlockStmt* block) {
   BlockStmt* instantiationPt = NULL;
 
-  if (block->parentExpr != NULL) {
-    // not a module or function level block
-  } else if (inMod && block == inMod->block) {
-    // module-level block
-  } else if (inFn != NULL) {
-    // TODO - probably remove this assert
-    INT_ASSERT(block->parentSymbol == inFn ||
-               isArgSymbol(block->parentSymbol) ||
-               isShadowVarSymbol(block->parentSymbol));
```

Comment:
I'm assuming this is the assert that was causing problems for you?  The TODO here does make me inclined to okay its removal, but I would be slacking in my review if I didn't at least double check why we think it is okay to remove it.

Reply:
Correct. Good point.

I made the change not because it was logically identical, but on semantics grounds. We care about the point of instantiation only for a function. (Even if it is a method on a generic class.) Therefore getVisibleFnsInstantiationPt() should return the POI only for its top-level blocks, such as `body` or `where`. So I wrote the code that does that, instead of going roundabout it.

Reply:
I forget, are we allowed to define nested functions at arbitrary points within a function?  Or do they have to be at the top level of the function body?

E.g. can I write this?
```chapel
proc foo(...) {
  if blah {
    proc innerFoo() { ... }
    innerFoo();
  }
}
```

Reply:
I don't think we can define classes in those bodies any more, so maybe it doesn't matter

Reply:
Yes, a nested function can be defined in a nested scope within the enclosing function. The relevant piece here is that a nested scope, such as the then-block of a conditional, does not have a POI. getVisibleFnsInstantiationPt() rightfully returns NULL in this case.

Re classes: a method on a generic class has its own POI when instantiated, it does not rely on the POI of the class. I think that classes don't have their own POIs.

Reply:
> Yes, a nested function can be defined in a nested scope within the enclosing function. The relevant piece here is that a nested scope, such as the then-block of a conditional, does not have a POI. getVisibleFnsInstantiationPt() rightfully returns NULL in this case.

That seems problematic, though.  If we've written a nested function that fulfills what a generic function is looking for, would we not be able to use the nested function?  Or would two calls at different scopes in the same function use the same instantiation, even if the first was able to see the nested function and the second wasn't?  E.g.

```chapel
proc genericFunc(param p) {
  poiCall();
}
proc other() {
  if blah {
    proc poiCall() { ... }
    genericFunc(1); // Should this instantiation use the definition above?
  }
  // I don't think this call should reuse the instantiation
  // since it's not in the scope of the poiCall definition
  genericFunc(1);
}
```

Reply:
Good example. To foreshadow the following discussion, the compiler does the right thing.

Let us see what is happening during resolution of the above code.

(A) When resolving either call to genericFunc(), POI does not come into play. getVisibleFunctionsImpl() goes up the lexical scopes until it reaches the body of the implicit module, where it finds the definition of genericFunc().

(B) The POI becomes relevant when resolving the call to poiCall() within genericFunc(). There, getVisibleFnsInstantiationPt() returns NULL whenever its argument is not the top-level block of genericFunc's body. For example, if the call to poiCall() were nested within a conditional:

```chpl
proc genericFunc(param p) {
  if blah {
    poiCall();
  }
}
```

then the first time getVisibleFnsInstantiationPt() is invoked it would return NULL. This is the correct behavior.

Meanwhile getVisibleFunctionsImpl() continues marching up the lexical scopes. Once it reaches genericFunc's top body block, getVisibleFnsInstantiationPt() wakes up and returns genericFunc's POI.

That POI itself can be nested or not. However, it will be the job of getVisibleFunctionsImpl() to traverse up the lexical scopes starting from whatever POI it got.

In the above example, the POI for the first call to genericFunc() is nested within the conditional whereas the second one is not. Either way, getVisibleFnsInstantiationPt() is not affected by this difference because it returns them without examining them.

<details>
<summary>A compilable example</summary>

```chpl
config var blah = true;

proc genericFunc(param p) {
  if blah {
    poiCall();
  }
}
proc other() {
  if blah {
    proc poiCall() { writeln("inner"); }
    genericFunc(1);
  }
  proc poiCall() { writeln("outer"); } // commenting this out would give an "unresolved" error
  genericFunc(1);
}

proc main {
  other();
}
```

</details>

Reply:
Okay, sweet!  Thanks for walking through it